converters/toZigbee.js

```diff
@@ -841,234 +842,93 @@ const converters = {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
-                }
-            } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('v')) {
-                newState.color = {h: value.h, s: value.s, v: value.v};
-                const hsv = utils.gammaCorrectHSV(utils.correctHue(value.h, meta), value.s, value.v);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
+
+                // Convert RGB to XY color mode because at the moment all known devices support this mode.
+                const xy = (newColor.rgb !== null) ? newColor.rgb.gammaCorrected().toXY() : newColor.xy;
+
+                // Some bulbs e.g. RB 185 C don't turn to red (they don't respond at all) when x: 0.701 and y: 0.299
+                // is send. These values are e.g. send by Home Assistant when clicking red in the color wheel.
+                // If we slighlty modify these values the bulb will respond.
+                // https://github.com/home-assistant/home-assistant/issues/31094
+                if (utils.getMetaValue(entity, meta.mapped, 'applyRedFix', 'allEqual', false) && xy.x == 0.701 && xy.y === 0.299) {
+                    xy.x = 0.7006;
+                    xy.y = 0.2993;
                 }
-            } else if (value.hasOwnProperty('hsv')) {
-                let hsv = value.hsv.split(',').map((i) => parseInt(i));
-                newState.color = {h: hsv[0], s: hsv[1], v: hsv[2]};
-                hsv = utils.gammaCorrectHSV(utils.correctHue(hsv[0], meta), hsv[1], hsv[2]);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
+
+                newState.color_mode = constants.colorMode[1];
+                newState.color = xy.toObject(4);
```

Comment:
When set via XY no precision rounding should be done (only in case of RGB)

Reply:
I added this because original `rgbToXY` function had this 4 decimal places rounding built in.

HA sends x/y rounded to 3 decimal places so its not an issue there, but with computed values we could get full 16 decimal places in state which I don't think are ever necessary.

I don't have strong opinion here, I can remove it if you prefer so.

Reply:
If with computed values you mean gamma corrected RGB I propose to only round in case of a gamma correction and don't do any rounding when x/y values were send.

Reply:
After reading: https://github.com/Koenkk/zigbee-herdsman-converters/pull/2453/files#r610569342

My opinion is that we should only round when computation has been done. If you directly set x/y or hue/saturation those exact values should be returned. Maybe the rounding can be done directly in the computation methods (e.g. .gammaCorrected())

Reply:
Addressed in https://github.com/Koenkk/zigbee-herdsman-converters/pull/2453/commits/3d3933ef5696c00d3b491e53f36ddae811e0e65c, I added another intermediate method `rounded()` to apply rounding and included if after `ColorXY` object is created from color corrected RGB input.